### DISCUSSION TITLE
Websocket client not connecting to the socket.io server

### DISCUSSION THREAD

**Discussion by hrishibawane** [socketio/socket.io#4299] [Q&A] [answered]
I have a basic socket.io server (version 4.4.1) to which I am trying to connect with a C++ websocket client. 
If I use ```?EIO=3&transport=websocket``` as query parameters in the URL, the connection is successful and the ```io.on("connection")``` event is triggered on the server.
However, if I use ```?EIO=4&transport=websocket```, the connection does get successful but the ```io.on("connection")``` event is not triggered on server.

I have even tried from Postman client using Raw Websocket client too. It shows the same behaviours with EIO=3 and EIO=4.
I understand that socket.io has additional handling over raw websocket connection, but when connecting with EIO=3 in the URL, there
 are no issues seen. I also understand that after v2, the breaking change was the reversal of Ping/Pong messages. However, that part comes later in the socket life cycle. This issue I am facing is concerning the point of connection at which different behaviours are seen with Engine.io versions.

Any pointers on what extra needs to be done on the client or server side to resolve this?

**Comment by hrishibawane** [comment, marked as answer]
So after debugging a lot I was able to figure this out.
The issue is that when using EIO v3 the server sends the CONNECT message packet. However, for EIO v4, the client has to send the CONNECT message packet.
CONNECT message packet is identified by text 40 where,
4 => engine.io packet type MESSAGE
0 => socket.io packet type CONNECT

This needs to be in the migration documentation along with the reversal of ping/pong messages.

**Reply by darrachequesne** [reply]
Hi! You can find more details about the Engine.IO & Socket.IO protocols there:

- https://github.com/socketio/engine.io-protocol
- https://github.com/socketio/socket.io-protocol

**Reply by Notadumdum** [reply]
@hrishibawane hey brother I have questions on this

**Comment by sergi** [comment]
I'm having this problem now. It's very confusing to me that `allowEIO3: true` doesn't prevent this. What's the point of that directive, if v2 clients can't connect to it anyway? Is there a way to prevent v4 to send the `40` on connect?

**Comment by Notadumdum** [comment]
Use a websocket client instead.. socket.io prevents you from successfully
connecting. Thats what i did..
